File: lib/features/activities/domain/entities/activity.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import '../../../../core/utils/datetime_utils.dart';

part 'activity.freezed.dart';

enum ActivityType {
  projectCreated,
  projectUpdated,
  projectDeleted,
  contentUploaded,
  summaryGenerated,
  querySubmitted,
  reportGenerated,
  memberAdded,
  memberRemoved,
}

@freezed
class Activity with _$Activity {
  const factory Activity({
    required String id,
    required String projectId,
    required ActivityType type,
    required String title,
    required String description,
    String? metadata,
    required DateTime timestamp,
    String? userId,
    String? userName,
  }) = _Activity;

  const Activity._();

  String get formattedTime {
    return DateTimeUtils.formatTimeAgo(timestamp);
  }

  String get icon {
    switch (type) {
      case ActivityType.projectCreated:
        return '📁';
      case ActivityType.projectUpdated:
        return '✏️';
      case ActivityType.projectDeleted:
        return '🗑️';
      case ActivityType.contentUploaded:
        return '📤';
      case ActivityType.summaryGenerated:
        return '✨';
      case ActivityType.querySubmitted:
        return '❓';
      case ActivityType.reportGenerated:
        return '📄';
      case ActivityType.memberAdded:
        return '👤';
      case ActivityType.memberRemoved:
        return '👥';
    }
  }
}

File: backend/models/activity.py
from sqlalchemy import Column, String, DateTime, ForeignKey, Text, Enum as SQLEnum
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
import uuid
from datetime import datetime
import enum

from db.database import Base

class ActivityType(enum.Enum):
    PROJECT_CREATED = "project_created"
    PROJECT_UPDATED = "project_updated"
    PROJECT_DELETED = "project_deleted"
    CONTENT_UPLOADED = "content_uploaded"
    SUMMARY_GENERATED = "summary_generated"
    QUERY_SUBMITTED = "query_submitted"
    REPORT_GENERATED = "report_generated"
    MEMBER_ADDED = "member_added"
    MEMBER_REMOVED = "member_removed"

class Activity(Base):
    __tablename__ = "activities"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    project_id = Column(UUID(as_uuid=True), ForeignKey("projects.id", ondelete="CASCADE"), nullable=False)
    type = Column(SQLEnum(ActivityType), nullable=False)
    title = Column(String(255), nullable=False)
    description = Column(Text, nullable=False)
    activity_metadata = Column(Text)
    timestamp = Column(DateTime, default=datetime.utcnow, nullable=False)
    user_id = Column(String(255))
    user_name = Column(String(255))
    
    # Relationships
    project = relationship("Project", back_populates="activities")
    
    def to_dict(self):
        return {
            "id": str(self.id),
            "project_id": str(self.project_id),
            "type": self.type.value,
            "title": self.title,
            "description": self.description,
            "metadata": self.activity_metadata,
            "timestamp": self.timestamp.isoformat(),
            "user_id": self.user_id,
            "user_name": self.user_name
        }